xbmc/network/httprequesthandler/HTTPPythonHandler.cpp

```diff
@@ -0,0 +1,242 @@
+  {
+    m_response.type = HTTPError;
+    if (m_response.status == MHD_HTTP_FOUND)
+      m_response.status = MHD_HTTP_INTERNAL_SERVER_ERROR;
+
+    return;
+  }
+
+  m_type = boost::dynamic_pointer_cast<ADDON::CWebinterface>(m_addon)->GetType();
+
+  m_response.type = HTTPMemoryDownloadNoFreeCopy;
+  m_response.status = MHD_HTTP_OK;
+  
+  // determine the last modified date
+  const CURL pathToUrl(m_scriptPath);
+  struct __stat64 statBuffer;
+  if (XFILE::CFile::Stat(pathToUrl, &statBuffer) != 0)
+    return;
+
+  struct tm* time;
+#ifdef HAVE_LOCALTIME_R
+  struct tm result = {};
+  time = localtime_r((time_t*)&statBuffer.st_mtime, &result);
+#else
+  time = localtime((time_t *)&statBuffer.st_mtime);
+#endif
+  if (time == NULL)
+    return;
+
+  m_lastModified = *time;
+}
+
+bool CHTTPPythonHandler::CanHandleRequest(const HTTPRequest &request)
+{
+  return URIUtils::GetExtension(request.pathUrl).compare(".py") == 0;
+}
+
+int CHTTPPythonHandler::HandleRequest()
+{
+  if (m_response.type == HTTPError)
+    return MHD_YES;
+
+  std::vector<std::string> args;
+  args.push_back(m_scriptPath);
+
+  try
+  {
+    HTTPPythonRequest* pythonRequest = new HTTPPythonRequest();
+    pythonRequest->connection = m_request.connection;
+    pythonRequest->file = URIUtils::GetFileName(m_request.pathUrl);
+    CWebServer::GetRequestHeaderValues(m_request.connection, MHD_GET_ARGUMENT_KIND, pythonRequest->getValues);
+    CWebServer::GetRequestHeaderValues(m_request.connection, MHD_HEADER_KIND, pythonRequest->headerValues);
+    pythonRequest->method = m_request.method;
+    pythonRequest->postValues = m_postFields;
+    pythonRequest->requestContent = m_requestData;
+    pythonRequest->responseType = HTTPNone;
+    pythonRequest->responseLength = 0;
+    pythonRequest->responseStatus = MHD_HTTP_OK;
+    pythonRequest->url = m_request.pathUrlFull;
+    pythonRequest->path = m_request.pathUrl;
+    pythonRequest->version = m_request.version;
+    pythonRequest->requestTime = CDateTime::GetCurrentDateTime();
+    pythonRequest->lastModifiedTime = m_lastModified;
+
+    std::string hostname;
+    uint16_t port;
+    if (GetHostnameAndPort(hostname, port))
+    {
+      pythonRequest->hostname = hostname;
+      pythonRequest->port = port;
+    }
+
+    CHTTPPythonInvoker* pythonInvoker = NULL;
+    if (m_type == ADDON::WebinterfaceTypeModPython)
+      pythonInvoker = new CHTTPModPythonInvoker(&g_pythonParser, pythonRequest);
+    else
+    {
+      CLog::Log(LOGWARNING, "WebServer: unable to run python script at %s with an unknown invoker", m_scriptPath.c_str());
+
+      m_response.status = MHD_HTTP_INTERNAL_SERVER_ERROR;
+      m_response.type = HTTPError;
+
+      return MHD_YES;
+    }
+
+    LanguageInvokerPtr languageInvokerPtr(pythonInvoker);
+    int result = CScriptInvocationManager::Get().ExecuteSync(m_scriptPath, languageInvokerPtr, m_addon, args, 30000, false);
```

Comment:
Uhm no:) But that you don't have to setup the interpreter for every request (for performance) seem to be one of the things mentioned as motivation in the spec. Though you aren't going to have to handle *that* many requests are you? What I meant was something simple like: https://docs.python.org/2/extending/extending.html#calling-python-functions-from-c

In WSGI one should **not** run the script but instead something like:

1. Do a regular import of the script using `PyImport_Import`
2. Get reference to the callable in that module with `PyObject_GetAttrString`
3. Invoke/get the response from that callable with `PyObject_CallObject`

That's the low level stuff. I have no idea where to put it. It replaces parts of `CPythonInvoker::Execute` but doesn't fit very well with the whole executing a script by file name concept.